### DISCUSSION TITLE
Different linker support

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#197] [Q&A] [answered]
I personally use [mold](https://github.com/rui314/mold) as my default linker on my Debian machine.

Since GCC 12 the flag `-fuse-ld=` can be used to choose a different linker if you like.

Is there a way to append `-fuse-ld=mold` to LDFLAG? If I'm not mistaken there's a way to modify the global settings via Lua script or something like that.

**Comment by edubart** [comment, marked as answer]
> Is there a way to append -fuse-ld=mold to LDFLAG?

Yes, use the following in your Nelua code:
```lua
## ldflags '-fuse-ld=mold'
```

Although using `cflags` will also work:

```lua
## cflags '-fuse-ld=mold'
```

Technically it's more correct to use `ldflags`, since this is a linkage flag. The main difference between using `ldflags` instead of `cflags`, is that `ldflags` options won't be used when Nelua attempts to compile object files and static libraries.

This can also be set via command line, just like `--cflags`, you can also use `--ldflags`.

> If I'm not mistaken there's a way to modify the global settings via Lua script or something like that.

Yes, creating a `~/.config/nelua/neluacfg.lua` like the following:

```lua
return {
  cflags='-pipe', -- pipe option may speedup GCC compilation by creating less temporary files
  ldflags='-fuse-ld=mold', -- mold linker may speedup GCC linking
}
```

> I personally use [mold](https://github.com/rui314/mold) as my default linker on my Debian machine.

What makes mold appealing is that it is able to link hundreds of object files very fast taking full use of processor cores. Though Nelua usually compiles everything into a single object file, and that is less parallelizable. I am not aware of mold internals, but I would expect to make less difference in this case. There is also another linker I use sometimes to link faster, the `gold` one, you can use with `-fuse-ld=gold`. But for most users, I recommend not changing the linker to avoid potential issues, I have run into some random issues when using non default linkers with GCC before.

**Reply by stefanos82** [reply]
Thank you for the suggestions; all of them look quite appealing.

Yes, I know about gold, because my Debian system uses it in place of `ld`.

Maybe it was installed by me in the past(?), I cannot remember; what matters is that I used to use it as my default linker, but now I use `mold`.

This was the motivation since day 1:

![image](https://user-images.githubusercontent.com/1777466/197247979-38427b8a-dd72-4cf4-a91c-9428031a9951.png)

Update: I forgot to mention that Rui is the same person that designed and implemented LLVM ld and mold!
